### PR TITLE
[Core] Support multi-node inference(eager and cuda graph)

### PR DESCRIPTION
**Pull request by esmeetu**
After #3625 , we can smoothly support multi-node inference. Thanks for previous @youkaichao great work!

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/parallel_utils/pynccl.py [hunk 1]
```diff
@@ -219,25 +220,22 @@ def __init__(
...
             self.unique_id = ncclGetUniqueId()
         else:
             self.unique_id = NcclUniqueId()
-        tensor = torch.ByteTensor(list(self.unique_id.internal)).cuda(
-            self.rank)
+        tensor = torch.ByteTensor(list(
+            self.unique_id.internal)).cuda(local_rank)
         dist.broadcast(tensor, src=0)
         byte_list = tensor.cpu().tolist()
-        self.unique_id = NcclUniqueId()
```
Comment (esmeetu) [inline vllm/model_executor/parallel_utils/pynccl.py:233]: This line seems useless, as `self.unique_id` has already defined in line 227.

Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py:233]: Seems good to me.

File: vllm/model_executor/parallel_utils/pynccl.py [hunk 2]
```diff
@@ -219,25 +220,22 @@ def __init__(
                                     store=store,
                                     group_name=group_name,
                                     pg_options=pg_options)
-        self.world_size = dist.get_world_size()
-        self.rank = dist.get_rank()
```
Comment (youkaichao) [inline vllm/model_executor/parallel_utils/pynccl.py:223]: We cound still save `self.world_size/self.rank/self.local_rank` , for convenience.

**Review by youkaichao — APPROVED** [review]
Overall it looks good to me. I thought to implement it in several days but had no bandwidth. Thanks for the fix!

cc @simon-mo can we have some basic multi-node multi-gpu correctness test? I feel this PR should be good in general, but have a test would be better.

**Comment by simon-mo** [comment]
We don't have one. Hard to setup one in our current CI. It would be useful if you can test manually for this PR just once.

**Comment by youkaichao** [comment]
@esmeetu did you test it under multi node setting? I currently don't have a working environment at hand.

**Comment by esmeetu** [comment]
> @esmeetu did you test it under multi node setting? I currently don't have a working environment at hand.

Of course, I have tested this at thesr two mode. And it's good for me.

**Comment by cadedaniel** [comment]
What's the plan for adding a test for this? Basically, we can't rely on untested code -- it will always break given enough time.

**Comment by youkaichao** [comment]
@cadedaniel do you have any suggestions for testing with multi-node? I suppose that should be not easy to set up.

**Comment by cadedaniel** [comment]
Yeah, needs some work in CI to support these kind of features

**Comment by esmeetu** [comment]
> What's the plan for adding a test for this? Basically, we can't rely on untested code -- it will always break given enough time.

You are right. I just finished e2e test which is not enough. Could we regard it as an **experimental** feature for 0.4.0? And as it doesn't break intra-node CI tests, i think it's acceptable.
For the multi-node test, i need some help to finish this. cc @simon-mo

**Pull request merged**